pkg/storegateway/chunkscache/cache.go

```diff
@@ -0,0 +1,145 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package chunkscache
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	gokit_log "github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/cache"
+	"github.com/oklog/ulid"
+	"github.com/opentracing/opentracing-go/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/prometheus/tsdb/chunks"
+
+	"github.com/grafana/mimir/pkg/util/spanlogger"
+)
+
+// Range is a contiguous range of chunks. Start is the ref of the first chunk and NumChunks is how many chunks
+// there are in the range.
+type Range struct {
+	BlockID   ulid.ULID
+	Start     chunks.ChunkRef
+	NumChunks int
+}
+
+type ChunksCache interface {
+	FetchMultiChunks(ctx context.Context, userID string, ranges []Range) (hits map[Range][]byte)
+	StoreChunks(ctx context.Context, userID string, r Range, v []byte)
+}
+
+type TracingCache struct {
+	c ChunksCache
+	l gokit_log.Logger
+}
+
+func NewTracingCache(c ChunksCache, l gokit_log.Logger) TracingCache {
+	return TracingCache{
+		c: c,
+		l: l,
+	}
+}
+
+func (c TracingCache) FetchMultiChunks(ctx context.Context, userID string, ranges []Range) (hits map[Range][]byte) {
+	l := spanlogger.FromContext(ctx, c.l)
+	hits = c.c.FetchMultiChunks(ctx, userID, ranges)
+
+	l.LogFields(
+		log.String("name", "ChunksCache.FetchMultiChunks"),
+		log.Int("ranges_requested", len(ranges)),
+		log.Int("ranges_hit", len(hits)),
+		log.Int("ranges_hit_bytes", hitsSize(hits)),
+		log.Int("ranges_misses", len(ranges)-len(hits)),
+	)
+	return
+}
+
+func hitsSize(hits map[Range][]byte) (size int) {
+	for _, b := range hits {
+		size += len(b)
+	}
+	return
+}
+
+func (c TracingCache) StoreChunks(ctx context.Context, userID string, r Range, v []byte) {
+	c.c.StoreChunks(ctx, userID, r, v)
+}
+
+const (
+	defaultTTL = 7 * 24 * time.Hour
+)
+
+// DskitChunksCache is a cache-based index cache.
```

Comment:
```suggestion
// DskitChunksCache is a cache-based chunks cache.
```

Reply:
I think that this comment doesn't make much sense to begin with :D How about
```suggestion
// DskitChunksCache is a ChunksCache which uses the dskit caching interface.
```